### ISSUE TITLE
Define TD for script manager Thing

### ISSUE THREAD

**Issue by zolkis** [w3c/wot-scripting-api#23]
As discussed in #22, a WoT Runtime will encapsulate special Things that may provide runtime and platform related functionality to external clients. One example is script management. Scripts can be used to create new Things on the local runtime and add new interactions.

The use cases (tentatively) are handling requests to:
- run a serialized script on the local runtime (e.g. to test it works)
- install a serialized script and return a script handle (e.g. UUID). This may have the following sub-steps:
  * verify the script and fail if invalid
  * save the script locally and define a handle. Persistency is needed for being able to run a script at boot.
  * optionally mark the script to run automatically at next boot with given sequence number
- run a script by handle (e.g. to test it works)
- stop a script by handle
- mark a script by handle to run at (re)boot with specified sequence number
- unmark a script by handle
- uninstall a script specified by handle
  * unmark the script in boot sequence
  * delete from persistent storage
  * remove the handle.

The goal is to define a Thing Description for such a script manager Thing. This issue serves for discussion and proposals.

The Thing could be described in WebIDL as follows (names are just for illustration).
```webidl
typedef ScriptID USVString;  // e.g. UUID

interface ScriptManagerThing {
  // main actions
  any run(USVString script);  // run a serialized script and return the result or error
  ScriptID? install(USVString script, optional unsigned long bootSequence);  // save[+mark+run]
  bool uninstall(ScriptID handle);  // uninstall a script by handle (stop, unmark, delete)

  // fine-grained actions
  bool start(ScriptID handle);  // start a saved script by handle
  bool stop(ScriptID handle);  // stop a running script by handle
  bool mark(ScriptId handle, unsigned long bootSequence);  // run at next boot
  bool unmark(ScriptId handle);
};
```
The `install` action, given a `serializedScript`, will verify and save the given script associated with a `handle`, then if  `bootSequence` was given, then run the `mark(handle, bootSequence)` and `start(handle)` actions.

@h0ru5 please check the use cases and WebIDL description.
@sebastiankb @mkovatsc @danicic @mariapoveda @vcharpenay please help translating to TD :).

**Comment by h0ru5** [comment]
I think in broad strokes the use cases & signatures fit.

For the TD, I think we should not standardize the names, but "@types" of actions.

Some minor remarks:
* mark/unmark: it's not necessary a boot, its basically "autostart" (i.e. it could be only the runtime / "agent" restarting)
* bootSequence: I guess for symmetry it should be optional also mark - but I did not understand what it does
* During my experiments, I found a `restartRuntime` (resp. reboot following the current terminology) very useful
* running a script produces state. It attaches handlers, starts `setInterval` etc. This cannot be garbage-collected. IIRC the lifecycle model is that scripts don't have an own main-loop but only attach handlers.  Stopping it could mean different things (just thinking out loud here, might be incomplete): 
** (a) just a "kill -9" for scripts that "hang"/loop forever - state remains
** (b) we send an event to the script which can be used to unregister handlers and clean up 
** (c) the runtime is expected to track those things "under the covers" and clean up

I do prefer (b) as we might want to have other runtime events later.
(you can still do (c) and after a timeout (a) - but I would absolutely not see this mandatory)

**Comment by zolkis** [comment]
Yes, I forgot the `reboot` or `restart` action, but it's not strictly a script management action. I added it anyway. Sometimes when a new Thing is added, the HW needs reboot, and sometimes a soft restart is enough. I used `restart` and it's up to the Thing, bindings, and runtime to figure out if a reboot is actually needed.
 
Also, made sequence optional on `mark`. If not provided, the script is queued last (where order matters).

True, scripts just attach handlers, but `stop` might be what the specific JS runtime implementations are capable of doing, which AFAIK is mostly stopping the script, which in our case means indeed mainly (b).

Buggy scripts can compromise a runtime even over reboots, but there are mechanisms to detect and break such cycles, so I assume implementations can encapsulate it. In the worst case the developer needs to re-flash, which would be the only possibility if the script manager Thing wasn't there.

**Comment by mmccool** [comment]
- Given the silmilarity to services (eg systemd) I think a similar terminology of "start", "stop", and "restart" makes sense.  However, names should probably be the last thing discussed
- It would be worth looking at similar things like AWS's "Lambda" system for a couple of reasons: 1 are there things they do that we should also consider?  2 is there a set of people using these other things that we don't want to confuse, and/or that we want to help them reuse their expertise?

**Comment by mmccool** [comment]
Are Things created by scripts like objects?  Should there be a kind of Factory pattern here: you load a script into the runtime, then "create" a Thing running that script, maybe stop and start it, maybe ultimately destroy it?

**Comment by mmccool** [comment]
Also, scripts probably should not be strings, but some kind of abstract type created by some kind of "packager".   Also, while we may not want to have a type system for different kinds of scripts, I think a runtime should be able to "reject" a script "package" if when it opens it it finds it can't run it for whatever reason.  It should be possible for installation to fail and to signal this.   I assume in the API given this could be signalled by returning a special "null" scriptID.

**Comment by zolkis** [comment]
How the Thing objects are created and maintained should also be part of the WoT Runtime.
In principle a script can create one or more Thing objects, all of which get integrated in the WoT Runtime mainloop for servicing requests.

"Serialized scripts" are indeed opaque bitstreams in principle, but we need to be able to pass them through HTTP, and for at least JavaScript they will be strings (USVString in WebIDL). It's an open for other than JS runtimes.

**Comment by mmccool** [comment]
I guess my main point is that we can't assume the "Scripts" will be JS.   Even if we only specify JS for now, we should define an open-ended package structure that allows some metadata about the payload (for instance, what language it is, and maybe what resources it needs to run) to be specified.   This could take the form (for example) of a very simple header containing name-value pairs like in HTTP.

**Comment by zolkis** [comment]
That is true, but the WoT Runtime will be different for each language. Therefore we have agreed to initially limit the Scripting API work scope to JavaScript and TypeScript, in order to have the chance to get something done. :).

When the clear need for other runtimes emerge, it will need to have 
- addendum to the Scripting spec and 
- addendum for TD for script management Thing
- WoT Runtime design recommendations doc, including all encapsulated details, such as script serialization etc.
- test suite.

**Comment by mmccool** [comment]
Created some separate issues for some of the points above.   I actually forgot I had previously brought them up :-P
